### ISSUE TITLE
Upsample bilinear 2d decomposition does not match native implementation for uint8

### ISSUE THREAD

**Issue by abulavin** [pytorch/pytorch#113024]
### 🐛 Describe the bug

The `torch.nn.functional.interpolate` op in its implementation will call `torch._C.nn.upsample_bilinear2d` when `mode=bilinear` and `input.dim() == 4`. However when the decomposition for `upsample_bilinear2d` is called, as opposed to the native implementation of `torch._C.nn.upsample_bilinear2d`, the results of `interpolate` differ.

After doing some digging, I can see the difference comes from the results of `torch._C.nn.upsample_bilinear2d` and its respective decomposition `torch._decomps.decompositions.upsample_bilinear2d_vec` not matching. Reproducer below
```python
import torch
from torch._decomp.decompositions import upsample_bilinear2d_vec

torch.manual_seed(42)

sample_input = torch.randint(0, 255, size=[2, 3, 270, 270], 
                            dtype=torch.uint8)
size = (130, 130)

native_result = torch._C._nn.upsample_bilinear2d(sample_input, 
                                                 size, False, None)
decomp_result = upsample_bilinear2d_vec(sample_input, 
                                        size, False, None)

assert torch.eq(native_result, decomp_result).all() # fails

```

This happens with the input is of dtype `uint8`. The expected behaviour I would like to see is that the decomposition produces the same output as the native implementation.

### Versions

```
PyTorch version: 2.1.0+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 16.0.6 (https://github.com/conda-forge/clangdev-feedstock 60e0ed07baadaf59c7646015aff34fc0352327f0)
CMake version: version 3.25.3
Libc version: glibc-2.35

Python version: 3.10.12 | packaged by conda-forge | (main, Jun 23 2023, 22:40:32) [GCC 12.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-57-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   43 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          256
On-line CPU(s) list:             0-255
Vendor ID:                       AuthenticAMD
Model name:                      AMD EPYC 7742 64-Core Processor
CPU family:                      23
Model:                           49
Thread(s) per core:              2
Core(s) per socket:              64
Socket(s):                       2
Stepping:                        0
BogoMIPS:                        4491.60
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca sme sev sev_es
Virtualization:                  AMD-V
L1d cache:                       4 MiB (128 instances)
L1i cache:                       4 MiB (128 instances)
L2 cache:                        64 MiB (128 instances)
L3 cache:                        512 MiB (32 instances)
NUMA node(s):                    8
NUMA node0 CPU(s):               0-15,128-143
NUMA node1 CPU(s):               16-31,144-159
NUMA node2 CPU(s):               32-47,160-175
NUMA node3 CPU(s):               48-63,176-191
NUMA node4 CPU(s):               64-79,192-207
NUMA node5 CPU(s):               80-95,208-223
NUMA node6 CPU(s):               96-111,224-239
NUMA node7 CPU(s):               112-127,240-255
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Mitigation; untrained return thunk; SMT enabled with STIBP protection
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.0
[pip3] torch==2.1.0+cpu
[pip3] torchvision==0.16.0+cpu
[conda] numpy                     1.26.0          py310hb13e2d6_0    conda-forge
[conda] torch                     2.1.0+cpu                pypi_0    pypi
[conda] torchvision               0.16.0+cpu               pypi_0    pypi
```

**Comment by mingfeima** [comment]
@Xia-Weiwen could you please take a look at this one?

**Comment by Xia-Weiwen** [comment]
Hi @abulavin Thank you for reporting the issue. I will investigate it.

**Comment by vfdev-5** [comment]
@abulavin pytorch nightly (e.g. 2.3.0.dev20240104+cu121) is already working "correctly" on uint8 input. Your repro code from gives max abs error = 1 which is ok due roundings errors for uint8. We have a similar error for aten op when compare: 1) pure uint8 implementation vs 2) uint8 input -> cast to float -> apply interpolation -> round -> cast back to uint8.

**Comment by lezcano** [comment]
While I agree that, in general, we cannot match bit-by-bit the result of eager ops, and thus castings float -> int are going to be problematic, we should mitigate these issues whenever possible.
